macOS: hid4java incompatible architecture on arm64

When trying to launch 0.12 rc3 which is a universal app, the app will start but will not connect to server. This is caused by slimevr.jar not starting:
```
/usr/bin/java -Xmx512M -jar /Volumes/slimevr\ 1/slimevr.app/Contents/MacOS/slimevr.jar  run
15:39:20 [INFO] Using log folder: /Users/Admin/Library/Logs/dev.slimevr.SlimeVR
15:39:20 [INFO] Running version v0.12.0-rc.3
15:39:20 [INFO] Using config dir: /Users/Admin/Library/Application Support/dev.slimevr.SlimeVR/vrconfig.yml
15:39:20 [INFO] Starting the tracker server on port 6969...
15:39:20 [SEVERE] SLF4J: No SLF4J providers were found.
15:39:20 [SEVERE] SLF4J: Defaulting to no-operation (NOP) logger implementation
15:39:20 [SEVERE] SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
15:39:21 [INFO] [WebSocket] Web Socket VR Bridge started on port 21110
15:39:21 [SEVERE] org.hid4java.HidException: Hidapi did not initialise: dlopen(/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp, 0x0009): tried: '/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp' (no such file), '/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64'))
15:39:21 [SEVERE] 	at org.hid4java.HidDeviceManager.<init>(HidDeviceManager.java:87)
15:39:21 [SEVERE] 	at org.hid4java.HidServices.<init>(HidServices.java:82)
15:39:21 [SEVERE] 	at org.hid4java.HidManager.getHidServices(HidManager.java:75)
15:39:21 [SEVERE] 	at dev.slimevr.desktop.tracking.trackers.hid.TrackersHID.<init>(TrackersHID.kt:36)
15:39:21 [SEVERE] 	at dev.slimevr.desktop.Main.main(Main.kt:130)
15:39:21 [SEVERE] Caused by: java.lang.UnsatisfiedLinkError: dlopen(/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp, 0x0009): tried: '/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64')), '/System/Volumes/Preboot/Cryptexes/OS/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp' (no such file), '/Users/Admin/Library/Caches/JNA/temp/jna16742166045332538375.tmp' (mach-o file, but is an incompatible architecture (have 'x86_64', need 'arm64e' or 'arm64'))
15:39:21 [SEVERE] 	at com.sun.jna.Native.open(Native Method)
15:39:21 [SEVERE] 	at com.sun.jna.NativeLibrary.loadLibrary(NativeLibrary.java:300)
15:39:21 [SEVERE] 	at com.sun.jna.NativeLibrary.getInstance(NativeLibrary.java:481)
15:39:21 [SEVERE] 	at com.sun.jna.Library$Handler.<init>(Library.java:197)
15:39:21 [SEVERE] 	at com.sun.jna.Native.load(Native.java:618)
15:39:21 [SEVERE] 	at com.sun.jna.Native.load(Native.java:592)
15:39:21 [SEVERE] 	at org.hid4java.jna.HidrawHidApiLibrary.<clinit>(HidrawHidApiLibrary.java:36)
15:39:21 [SEVERE] 	at org.hid4java.jna.HidApi.init(HidApi.java:105)
15:39:21 [SEVERE] 	at org.hid4java.
```
Apparently updating JNA to 5.7.0 will fix this. See https://github.com/java-native-access/jna/blob/master/CHANGES.md#release-570

--- a/flake.nix
+++ b/flake.nix
@@ -96,6 +96,7 @@
               expat
               libayatana-appindicator
               libusb1
+              libudev-zero
             ])
             ++ lib.optionals pkgs.stdenv.isDarwin [
               pkgs.darwin.apple_sdk.frameworks.Security

--- a/server/desktop/build.gradle.kts
+++ b/server/desktop/build.gradle.kts
@@ -59,7 +59,7 @@ dependencies {
 	implementation("net.java.dev.jna:jna:5.+")
 	implementation("net.java.dev.jna:jna-platform:5.+")
 	implementation("com.fazecast:jSerialComm:2.10.2")
-	implementation("org.hid4java:hid4java:0.7.0")
+	implementation("org.hid4java:hid4java:0.8.0")
 }
 
 tasks.shadowJar {

--- a/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
+++ b/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
@@ -245,7 +245,7 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 	}
 
 	override fun hidDataReceived(p0: HidServicesEvent?) {
-		// TODO: Use this instead of TrackersHID#dataRead
+		// Seems to have issues with the size of data returned
 	}
 
 	companion object {

--- a/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
+++ b/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
@@ -32,7 +32,9 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 	private val hidServices: HidServices
 
 	init {
-		hidServicesSpecification.isAutoStart = false
+		hidServicesSpecification.setAutoStart(false)
+		hidServicesSpecification.setAutoDataRead(true)
+		hidServicesSpecification.setDataReadInterval(0)
 		hidServices = HidManager.getHidServices(hidServicesSpecification)
 		hidServices.addHidServicesListener(this)
 		val dataReadThread = Thread(dataReadRunnable)
@@ -43,7 +45,7 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 
 	private fun checkConfigureDevice(hidDevice: HidDevice) {
 		if (hidDevice.vendorId == 0x2FE3 && hidDevice.productId == 0x5652) { // TODO: Use correct ids
-			if (!hidDevice.isOpen) {
+			if (hidDevice.isClosed) {
 				check(hidDevice.open()) { "Unable to open device" }
 			}
 			// TODO: Configure the device here
@@ -148,83 +150,7 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 				return@Runnable
 			}
 			hidServices.start()
-			while (true) {
-				try {
-					sleep(0) // Possible performance impact
-				} catch (e: InterruptedException) {
-					currentThread().interrupt()
-					break
-				}
-				dataRead() // not in try catch?
-			}
 		}
-
-	private fun dataRead() {
-		synchronized(devicesByHID) {
-			var devicesPresent = false
-			for ((hidDevice, deviceList) in devicesByHID) {
-				val dataReceived: Array<Byte> = try {
-					hidDevice.read(80, 1) // Read up to 80 bytes, timeout 1ms
-				} catch (e: NegativeArraySizeException) {
-					continue // Skip devices with read error (Maybe disconnected)
-				}
-				devicesPresent = true // Even if the device has no data
-				if (dataReceived.isNotEmpty()) {
-					// Process data
-					// TODO: make this less bad
-					if (dataReceived.size % 20 != 0) {
-						LogManager.info("[TrackerServer] Malformed HID packet, ignoring")
-					}
-					val packetCount = dataReceived.size / 20
-					var i = 0
-					while (i < packetCount * 20) {
-						// dataReceived[i] //for later
-						val idCombination = dataReceived[i + 1].toInt()
-						val rssi = -dataReceived[i + 2].toInt()
-						val battery = dataReceived[i + 3].toInt()
-						val battery_mV = dataReceived[i + 5].toInt() and 255 shl 8 or (dataReceived[i + 4].toInt() and 255)
-						val q = floatArrayOf(0f, 0f, 0f, 0f)
-						val a = floatArrayOf(0f, 0f, 0f)
-						for (j in 0..3) { // quat received as fixed 14
-							var buf =
-								dataReceived[i + 6 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 6 + j * 2].toInt() and 255)
-							buf -= 32768 // uint to int
-							q[j] = buf / (1 shl 14).toFloat() // fixed 14 to float
-						}
-						for (j in 0..2) { // accel received as fixed 7, in m/s
-							var buf =
-								dataReceived[i + 14 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 14 + j * 2].toInt() and 255)
-							buf -= 32768 // uint to int
-							a[j] = buf / (1 shl 7).toFloat() // fixed 7 to float
-						}
-						val trackerId = idCombination and 0b1111
-						val deviceId = (idCombination shr 4) and 0b1111
-						val device = deviceIdLookup(hidDevice, deviceId, deviceList)
-						// server wants tracker to be unique, so use combination of hid serial and full id
-						setUpSensor(device, trackerId, IMUType.UNKNOWN, TrackerStatus.OK)
-						val tracker = device.getTracker(trackerId)!!
-
-						tracker.signalStrength = rssi
-						// tracker.batteryVoltage = if (battery and 128 == 128) 5f else 0f // Charge status
-						tracker.batteryVoltage = battery_mV.toFloat() * 0.001f
-						tracker.batteryLevel = (battery and 127).toFloat()
-						var rot = Quaternion(q[0], q[1], q[2], q[3])
-						rot = AXES_OFFSET.times(rot)
-						tracker.setRotation(rot)
-						// TODO: I think the acceleration is wrong???
-						var acceleration = Vector3(a[0], a[1], a[2])
-						tracker.setAcceleration(acceleration)
-						tracker.dataTick()
-						i += 20
-					}
-					// LogManager.info("[TrackerServer] HID received $packetCount tracker packets")
-				}
-			}
-			if (!devicesPresent) {
-				sleep(10) // No hid device, "empty loop" so sleep to save the poor cpu
-			}
-		}
-	}
 
 	override fun run() { // Doesn't seem to run
 	}
@@ -244,7 +170,57 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 	}
 
 	override fun hidDataReceived(p0: HidServicesEvent?) {
-		// TODO: Use this instead of TrackersHID#dataRead
+		synchronized(devicesByHID) {
+			val hidDevice: HidDevice = p0?.getHidDevice()!!
+			val dataReceived: ByteArray = p0.getDataReceived()
+			// Process data
+			// TODO: make this less bad
+			if (dataReceived.size % 20 != 0) {
+				LogManager.info("[TrackerServer] Malformed HID packet (${dataReceived.size}), ignoring")
+			}
+			val packetCount = dataReceived.size / 20
+			var i = 0
+			while (i < packetCount * 20) {
+				// dataReceived[i] //for later
+				val idCombination = dataReceived[i + 1].toInt()
+				val rssi = -dataReceived[i + 2].toInt()
+				val battery = dataReceived[i + 3].toInt()
+				val battery_mV = dataReceived[i + 5].toInt() and 255 shl 8 or (dataReceived[i + 4].toInt() and 255)
+				val q = floatArrayOf(0f, 0f, 0f, 0f)
+				val a = floatArrayOf(0f, 0f, 0f)
+				for (j in 0..3) { // quat received as fixed 14
+					var buf =
+						dataReceived[i + 6 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 6 + j * 2].toInt() and 255)
+					buf -= 32768 // uint to int
+					q[j] = buf / (1 shl 14).toFloat() // fixed 14 to float
+				}
+				for (j in 0..2) { // accel received as fixed 7, in m/s
+					var buf =
+						dataReceived[i + 14 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 14 + j * 2].toInt() and 255)
+					buf -= 32768 // uint to int
+					a[j] = buf / (1 shl 7).toFloat() // fixed 7 to float
+				}
+				val trackerId = idCombination and 0b1111
+				val deviceId = (idCombination shr 4) and 0b1111
+				val device = deviceIdLookup(hidDevice, deviceId, devicesByHID.get(hidDevice)!!)
+				// server wants tracker to be unique, so use combination of hid serial and full id
+				setUpSensor(device, trackerId, IMUType.UNKNOWN, TrackerStatus.OK)
+				val tracker = device.getTracker(trackerId)!!
+				tracker.signalStrength = rssi
+				// tracker.batteryVoltage = if (battery and 128 == 128) 5f else 0f // Charge status
+				tracker.batteryVoltage = battery_mV.toFloat() * 0.001f
+				tracker.batteryLevel = (battery and 127).toFloat()
+				var rot = Quaternion(q[0], q[1], q[2], q[3])
+				rot = AXES_OFFSET.times(rot)
+				tracker.setRotation(rot)
+				// TODO: I think the acceleration is wrong???
+				var acceleration = Vector3(a[0], a[1], a[2])
+				tracker.setAcceleration(acceleration)
+				tracker.dataTick()
+				i += 20
+			}
+			// LogManager.info("[TrackerServer] HID received $packetCount tracker packets")
+		}
 	}
 
 	companion object {

--- a/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
+++ b/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
@@ -33,8 +33,6 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 
 	init {
 		hidServicesSpecification.setAutoStart(false)
-		hidServicesSpecification.setAutoDataRead(true)
-		hidServicesSpecification.setDataReadInterval(0)
 		hidServices = HidManager.getHidServices(hidServicesSpecification)
 		hidServices.addHidServicesListener(this)
 		val dataReadThread = Thread(dataReadRunnable)
@@ -150,7 +148,84 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 				return@Runnable
 			}
 			hidServices.start()
+			while (true) {
+				try {
+					sleep(0) // Possible performance impact
+				} catch (e: InterruptedException) {
+					currentThread().interrupt()
+					break
+				}
+				dataRead() // not in try catch?
+			}
 		}
+
+	private fun dataRead() {
+		synchronized(devicesByHID) {
+			var devicesPresent = false
+			for ((hidDevice, deviceList) in devicesByHID) {
+				val dataReceived: Array<Byte> = try {
+					hidDevice.read(80, 0) // Read up to 80 bytes
+				} catch (e: NegativeArraySizeException) {
+					continue // Skip devices with read error (Maybe disconnected)
+				}
+				devicesPresent = true // Even if the device has no data
+				if (dataReceived.isNotEmpty()) {
+					// Process data
+					// TODO: make this less bad
+					if (dataReceived.size % 20 != 0) {
+						LogManager.info("[TrackerServer] Malformed HID packet, ignoring")
+						continue // Don't continue with this data
+					}
+					val packetCount = dataReceived.size / 20
+					var i = 0
+					while (i < packetCount * 20) {
+						// dataReceived[i] //for later
+						val idCombination = dataReceived[i + 1].toInt()
+						val rssi = -dataReceived[i + 2].toInt()
+						val battery = dataReceived[i + 3].toInt()
+						val battery_mV = dataReceived[i + 5].toInt() and 255 shl 8 or (dataReceived[i + 4].toInt() and 255)
+						val q = floatArrayOf(0f, 0f, 0f, 0f)
+						val a = floatArrayOf(0f, 0f, 0f)
+						for (j in 0..3) { // quat received as fixed 14
+							var buf =
+								dataReceived[i + 6 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 6 + j * 2].toInt() and 255)
+							buf -= 32768 // uint to int
+							q[j] = buf / (1 shl 14).toFloat() // fixed 14 to float
+						}
+						for (j in 0..2) { // accel received as fixed 7, in m/s
+							var buf =
+								dataReceived[i + 14 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 14 + j * 2].toInt() and 255)
+							buf -= 32768 // uint to int
+							a[j] = buf / (1 shl 7).toFloat() // fixed 7 to float
+						}
+						val trackerId = idCombination and 0b1111
+						val deviceId = (idCombination shr 4) and 0b1111
+						val device = deviceIdLookup(hidDevice, deviceId, deviceList)
+						// server wants tracker to be unique, so use combination of hid serial and full id
+						setUpSensor(device, trackerId, IMUType.UNKNOWN, TrackerStatus.OK)
+						val tracker = device.getTracker(trackerId)!!
+
+						tracker.signalStrength = rssi
+						// tracker.batteryVoltage = if (battery and 128 == 128) 5f else 0f // Charge status
+						tracker.batteryVoltage = battery_mV.toFloat() * 0.001f
+						tracker.batteryLevel = (battery and 127).toFloat()
+						var rot = Quaternion(q[0], q[1], q[2], q[3])
+						rot = AXES_OFFSET.times(rot)
+						tracker.setRotation(rot)
+						// TODO: I think the acceleration is wrong???
+						var acceleration = Vector3(a[0], a[1], a[2])
+						tracker.setAcceleration(acceleration)
+						tracker.dataTick()
+						i += 20
+					}
+					// LogManager.info("[TrackerServer] HID received $packetCount tracker packets")
+				}
+			}
+			if (!devicesPresent) {
+				sleep(10) // No hid device, "empty loop" so sleep to save the poor cpu
+			}
+		}
+	}
 
 	override fun run() { // Doesn't seem to run
 	}
@@ -170,57 +245,7 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 	}
 
 	override fun hidDataReceived(p0: HidServicesEvent?) {
-		synchronized(devicesByHID) {
-			val hidDevice: HidDevice = p0?.getHidDevice()!!
-			val dataReceived: ByteArray = p0.getDataReceived()
-			// Process data
-			// TODO: make this less bad
-			if (dataReceived.size % 20 != 0) {
-				LogManager.info("[TrackerServer] Malformed HID packet (${dataReceived.size}), ignoring")
-			}
-			val packetCount = dataReceived.size / 20
-			var i = 0
-			while (i < packetCount * 20) {
-				// dataReceived[i] //for later
-				val idCombination = dataReceived[i + 1].toInt()
-				val rssi = -dataReceived[i + 2].toInt()
-				val battery = dataReceived[i + 3].toInt()
-				val battery_mV = dataReceived[i + 5].toInt() and 255 shl 8 or (dataReceived[i + 4].toInt() and 255)
-				val q = floatArrayOf(0f, 0f, 0f, 0f)
-				val a = floatArrayOf(0f, 0f, 0f)
-				for (j in 0..3) { // quat received as fixed 14
-					var buf =
-						dataReceived[i + 6 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 6 + j * 2].toInt() and 255)
-					buf -= 32768 // uint to int
-					q[j] = buf / (1 shl 14).toFloat() // fixed 14 to float
-				}
-				for (j in 0..2) { // accel received as fixed 7, in m/s
-					var buf =
-						dataReceived[i + 14 + j * 2 + 1].toInt() and 255 shl 8 or (dataReceived[i + 14 + j * 2].toInt() and 255)
-					buf -= 32768 // uint to int
-					a[j] = buf / (1 shl 7).toFloat() // fixed 7 to float
-				}
-				val trackerId = idCombination and 0b1111
-				val deviceId = (idCombination shr 4) and 0b1111
-				val device = deviceIdLookup(hidDevice, deviceId, devicesByHID.get(hidDevice)!!)
-				// server wants tracker to be unique, so use combination of hid serial and full id
-				setUpSensor(device, trackerId, IMUType.UNKNOWN, TrackerStatus.OK)
-				val tracker = device.getTracker(trackerId)!!
-				tracker.signalStrength = rssi
-				// tracker.batteryVoltage = if (battery and 128 == 128) 5f else 0f // Charge status
-				tracker.batteryVoltage = battery_mV.toFloat() * 0.001f
-				tracker.batteryLevel = (battery and 127).toFloat()
-				var rot = Quaternion(q[0], q[1], q[2], q[3])
-				rot = AXES_OFFSET.times(rot)
-				tracker.setRotation(rot)
-				// TODO: I think the acceleration is wrong???
-				var acceleration = Vector3(a[0], a[1], a[2])
-				tracker.setAcceleration(acceleration)
-				tracker.dataTick()
-				i += 20
-			}
-			// LogManager.info("[TrackerServer] HID received $packetCount tracker packets")
-		}
+		// TODO: Use this instead of TrackersHID#dataRead
 	}
 
 	companion object {

--- a/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
+++ b/server/desktop/src/main/java/dev/slimevr/desktop/tracking/trackers/hid/TrackersHID.kt
@@ -243,6 +243,10 @@ class TrackersHID(name: String, private val trackersConsumer: Consumer<Tracker>)
 		// TODO:
 	}
 
+	override fun hidDataReceived(p0: HidServicesEvent?) {
+		// TODO: Use this instead of TrackersHID#dataRead
+	}
+
 	companion object {
 		/**
 		 * Change between IMU axes and OpenGL/SteamVR axes

